contracts/challenge-manager.sol

```diff
@@ -0,0 +1,172 @@
+pragma solidity 0.8.9;
+// SPDX-License-Identifier: UNLICENSED
+import './merkle-utils.sol';
+
+enum ChallengeStatus {
+    InProgress,
+    FraudDetected,
+    Forfeited
+}
+
+contract ChallengeManager {
+    bytes32 root;
+    uint256 splitFactor;
```

Comment:
If you have many variables that you would like to "store", it can make sense to think about packing them all inside one 32 byte variable. This is only about saving gas, mind -- so if that's not a concern yet it might not be the right time to optimize.

You have far more than 32 bytes here, so I would have a think about whether you need all of that storage in the first place (do you need 256 bit indices? does `mover` need to be of dynamic length?). Here's [an example](https://github.com/statechannels/statechannels/blob/2a30056ca78fa3d76b11a036a891f52814846e40/packages/nitro-protocol/contracts/StatusManager.sol#L37) of where we pack (and [unpack](https://github.com/statechannels/statechannels/blob/2a30056ca78fa3d76b11a036a891f52814846e40/packages/nitro-protocol/contracts/StatusManager.sol#L75)) multiple variables into one slot. It's possible you might be able to pack `consensusIndex,disputedIndex,lastMover` into 32 bytes if you restricted their types a bit more (I don't have enough context about these variables to know if that is safe). 

If you really need more than 32 bytes of storage, you could store a hash of the variables instead. The advantage is that you can then continue to only pay for make **one** SSTORE opcode (20K gas, by far the most expensive). The savings are likely going to outweigh the cost of packing and computing the hash.  But you will need to ensure clients have enough information to reconstruct the preimage of the hash so they can pass it in to the methods that need it. Those methods then rehash the preimage and check it against storage, and from that point on can trust the preimage and use it in computations. To help the clients track the preimage, you will probably want to emit an event (they cannot read from the contract storage). 

To decide on the correct approach you will probably want to audit all of your storage variables. For example, as far as I can tell, `splitFactor` is assigned in the constructor and then never updated (it is read, though). That would make it a good candidate for being an [immutable variable.](https://ethereum.stackexchange.com/questions/82240/what-is-the-immutable-keyword-in-solidity).

Reply:
@geoknee The immutable variable trick is a neat one. Thank you for pointing that out.